src/boot/efi/util.c

```diff
@@ -265,26 +265,36 @@ char16_t *xstr8_to_path(const char *str8) {
         return path;
 }
 
-void mangle_stub_cmdline(char16_t *cmdline) {
-        char16_t *p = cmdline;
+static bool shall_be_whitespace(char16_t c) {
+        return c <= 0x20U; /* All control characters + space */
+}
+
+char16_t* mangle_stub_cmdline(char16_t *cmdline) {
+        char16_t *p, *q, *e;
 
         if (!cmdline)
-                return;
+                return cmdline;
 
-        for (; *cmdline != '\0'; cmdline++)
-                /* Convert ASCII control characters to spaces. */
-                if (*cmdline <= 0x1F)
-                        *cmdline = ' ';
+        p = q = cmdline;
 
-        /* chomp the trailing whitespaces */
-        while (cmdline != p) {
-                --cmdline;
+        /* Skip initial whitespace */
+        while (shall_be_whitespace(*p))
+                p++;
 
-                if (*cmdline != ' ')
-                        break;
+        /* Turn inner control characters into proper spaces */
+        for (e = p; *p != 0; p++) {
+                if (shall_be_whitespace(*p)) {
+                        *(q++) = ' ';
+                        continue;
+                }
 
-                *cmdline = '\0';
+                *(q++) = *p;
+                e = q; /* remember last non-whitespace char */
         }
+
+        /* Chop off trailing whitespace */
+        *e = 0;
```

Comment:
It is not necessary to memory `q`, but simply assigning `*q = 0` at the end is enough.

Reply:
no, then we wouldn#t chop off trailing whitespace. `e` is only updated whenever we see a non-whitespace char, hence using that to chop off will work. q otoh will possibly point to a suffix of whitespace.